// @krassowski/jupyterlab-lsp/lib/features/highlights.js
import { DocumentHighlightKind } from '../lsp';
import { uris_equal } from '../utils';
import { FeatureSettings } from '../feature';
import { CodeMirrorIntegration } from '../editor_integration/codemirror';
import { ILSPFeatureManager, PLUGIN_ID } from '../tokens';
import { ISettingRegistry } from '@jupyterlab/settingregistry';
import { LabIcon } from '@jupyterlab/ui-components';
import highlightSvg from '../../style/icons/highlight.svg';
import highlightTypeSvg from '../../style/icons/highlight-type.svg';
export const highlightIcon = new LabIcon({
    name: 'lsp:highlight',
    svgstr: highlightSvg
});
export const highlightTypeIcon = new LabIcon({
    name: 'lsp:highlight-type',
    svgstr: highlightTypeSvg
});
const COMMANDS = [
    {
        id: 'highlight-references',
        execute: ({ connection, virtual_position, document }) => connection.getReferences(virtual_position, document.document_info),
        is_enabled: ({ connection }) => connection.isReferencesSupported(),
        label: 'Highlight references',
        icon: highlightIcon
    },
    {
        id: 'highlight-type-definition',
        execute: ({ connection, virtual_position, document }) => connection.getTypeDefinition(virtual_position, document.document_info),
        is_enabled: ({ connection }) => connection.isTypeDefinitionSupported(),
        label: 'Highlight type definition',
        icon: highlightTypeIcon
    }
];
export class HighlightsCM extends CodeMirrorIntegration {
    constructor() {
        super(...arguments);
        this.highlight_markers = [];
        this.handleHighlight = (items, documentUri) => {
            if (!uris_equal(documentUri, this.virtual_document.document_info.uri)) {
                return;
            }
            this.clear_markers();
            if (!items) {
                return;
            }
            for (let item of items) {
                let range = this.range_to_editor_range(item.range);
                let kind_class = item.kind
                    ? 'cm-lsp-highlight-' + DocumentHighlightKind[item.kind]
                    : '';
                let marker = this.highlight_range(range, 'cm-lsp-highlight ' + kind_class);
                this.highlight_markers.push(marker);
            }
        };
        this.onCursorActivity = async () => {
            var _a, _b;
            if (!((_b = (_a = this.virtual_editor) === null || _a === void 0 ? void 0 : _a.virtual_document) === null || _b === void 0 ? void 0 : _b.document_info)) {
                return;
            }
            let root_position;
            await this.virtual_editor.virtual_document.update_manager.update_done;
            try {
                root_position = this.virtual_editor
                    .getDoc()
                    .getCursor('start');
            }
            catch (err) {
                console.warn('LSP: no root position available');
                return;
            }
            let document;
            try {
                document = this.virtual_editor.document_at_root_position(root_position);
            }
            catch (e) {
                console.warn('LSP: Could not obtain virtual document from position', root_position);
                return;
            }
            if (document !== this.virtual_document) {
                return;
            }
            try {
                let virtual_position = this.virtual_editor.root_position_to_virtual_position(root_position);
                const highlights = await this.connection.getDocumentHighlights(virtual_position, this.virtual_document.document_info, false);
                if (!this.virtual_document.isDisposed) {
                    this.handleHighlight(highlights, this.virtual_document.document_info.uri);
                }
            }
            catch (e) {
                console.warn('Could not get highlights:', e);
            }
        };
    }
    register() {
        this.editor_handlers.set('cursorActivity', this.onCursorActivity);
        super.register();
    }
    remove() {
        this.handleHighlight = null;
        this.onCursorActivity = null;
        this.clear_markers();
        super.remove();
    }
    clear_markers() {
        for (let marker of this.highlight_markers) {
            marker.clear();
        }
        this.highlight_markers = [];
    }
}
const FEATURE_ID = PLUGIN_ID + ':highlights';
export const HIGHLIGHTS_PLUGIN = {
    id: FEATURE_ID,
    requires: [ILSPFeatureManager, ISettingRegistry],
    autoStart: true,
    activate: (app, featureManager, settingRegistry) => {
        const settings = new FeatureSettings(settingRegistry, FEATURE_ID);
        featureManager.register({
            feature: {
                editorIntegrationFactory: new Map([['CodeMirrorEditor', HighlightsCM]]),
                id: FEATURE_ID,
                name: 'LSP Highlights',
                settings: settings,
                commands: COMMANDS
            }
        });
    }
};
